controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,144 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// NamespaceReconciler reconciles the namespaces starting with openshift-*
+// nolint:revive
+type NamespaceReconciler struct {
+	client.Client
+	ctx    context.Context
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=ocs.openshift.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;update;patch
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// i.e,the annotations of the namespace
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	prevLogger := r.Log
+	defer func() { r.Log = prevLogger }()
+	r.Log = r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	r.ctx = ctx
+
+	r.Log.Info("Reconciling Namespaces starting with openshift-*.", "Namespace", klog.KRef(request.Namespace, request.Name))
+
+	ns := &corev1.NamespaceList{}
+
+	// Get all namespaces
+	err := r.List(ctx, ns)
+	if err != nil {
+		// Failed to list all the namespaces in the cluster
+		r.Log.Info("Re-fetching all the namespaces.")
+		return reconcile.Result{}, err
+	}
+
+	// Filter namespaces starting with openshift-*
+	opNS := getOpenshiftNamespaces(ns.Items)
```

Comment:
IMO we should not get the list of all namespaces and try to annotate all of them together. We should just add annotation only on the one namespace. 

We should use predicates to filter the openshift-* namespaces in the `SetupWithManager` func. 

@umangachapagain What is your opinion ?

Also, I  don't remember if we talked about restarting the CSI addons. I don't think it is a good idea to restart it. CSI-addons should pick up the changes automatically. @Madhu-1 WDYT?

Reply:
I am not sure how we can handle it using predicates here. AFAIU, predicates monitors create,delete,update and generic events and in this case when the controller will be deployed the namespaces that we want to annotate are already created and no new events will be triggered by them, so how will we be getting an event for each namespace such that we could annotate them?  Please correct me if I am misunderstanding anything here. This is the [ref](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) I checked out.

Regarding the csiaddons pod, this was advised by Madhu himself based on our last discussion and the JIRA ticket, Engineering response.

Reply:
Updated the code to use predicates and annotation to only one namespace at a time.
However, I am not sure where could we make use of the `sync.Once` type in the code now, I don't see any repetitive action happening now 🤔 . Could you please help me out with it?

Reply:
Lets meet and sort this.

Reply:
Updated the code as per our discussion, PTAL.